From b9d5f1a7cd111f0d46aab1ad98544757db964e5f Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 30 Jul 2024 09:47:58 +0200
Subject: [PATCH 1055/1695] drivers: stm32_hpdma: do not require the presence
 of st,protreg

Remove the constraint that made st,protreg mandatory as there are
use-cases where the driver can be used without setting a RIF
configuration such as when the CPU running OP-TEE is not TDCID.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f4dd57c2351ca25f4aa9c67e4cc118398c5cd4e8 drivers: stm32_hpdma: do not require the presence of st,protreg]
Change-Id: I49881c4c16f10ec8d4d57af0ab69f7d4519d60ef
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/395787
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_hpdma.c | 95 ++++++++++++++++++++++----------------
 1 file changed, 56 insertions(+), 39 deletions(-)

diff --git a/core/drivers/stm32_hpdma.c b/core/drivers/stm32_hpdma.c
index 89aa402be..1f7d52e74 100644
--- a/core/drivers/stm32_hpdma.c
+++ b/core/drivers/stm32_hpdma.c
@@ -77,7 +77,7 @@
 
 struct hpdma_pdata {
 	struct clk *hpdma_clock;
-	struct rif_conf_data conf_data;
+	struct rif_conf_data *conf_data;
 	uint32_t syscfg_arcr[2];
 	unsigned int nb_channels;
 	vaddr_t base;
@@ -123,8 +123,11 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 	uint32_t cidcfgr = 0;
 	unsigned int i = 0;
 
+	if (!hpdma_d->conf_data)
+		return TEE_SUCCESS;
+
 	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
-		if (!(BIT(i) & hpdma_d->conf_data.access_mask[0]))
+		if (!(BIT(i) & hpdma_d->conf_data->access_mask[0]))
 			continue;
 		/*
 		 * When TDCID, OP-TEE should be the one to set the CID filtering
@@ -153,23 +156,23 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 	}
 	/* Security and privilege RIF configuration */
 	io_clrsetbits32(hpdma_d->base + _HPDMA_PRIVCFGR, _HPDMA_PRIVCFGR_MASK &
-			hpdma_d->conf_data.access_mask[0],
-			hpdma_d->conf_data.priv_conf[0]);
+			hpdma_d->conf_data->access_mask[0],
+			hpdma_d->conf_data->priv_conf[0]);
 	io_clrsetbits32(hpdma_d->base + _HPDMA_SECCFGR, _HPDMA_SECCFGR_MASK &
-			hpdma_d->conf_data.access_mask[0],
-			hpdma_d->conf_data.sec_conf[0]);
+			hpdma_d->conf_data->access_mask[0],
+			hpdma_d->conf_data->sec_conf[0]);
 
 	/* Skip CID/semaphore configuration if not in TDCID state. */
 	if (!is_tdcid)
 		goto end;
 
 	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
-		if (!(BIT(i) & hpdma_d->conf_data.access_mask[0]))
+		if (!(BIT(i) & hpdma_d->conf_data->access_mask[0]))
 			continue;
 
 		io_clrsetbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
 				_HPDMA_CIDCFGR_CONF_MASK,
-				hpdma_d->conf_data.cid_confs[i]);
+				hpdma_d->conf_data->cid_confs[i]);
 
 		cidcfgr = io_read32(hpdma_d->base + _HPDMA_CIDCFGR(i));
 
@@ -204,21 +207,21 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 	 * next reset.
 	 */
 	io_clrsetbits32(hpdma_d->base + _HPDMA_RCFGLOCKR, _HPDMA_RCFGLOCKR_MASK,
-			hpdma_d->conf_data.lock_conf[0]);
+			hpdma_d->conf_data->lock_conf[0]);
 
 end:
 	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
 		/* Check that RIF config are applied, panic otherwise */
 		if ((io_read32(hpdma_d->base + _HPDMA_PRIVCFGR) &
-		     hpdma_d->conf_data.access_mask[0]) !=
-		    hpdma_d->conf_data.priv_conf[0]) {
+		     hpdma_d->conf_data->access_mask[0]) !=
+		    hpdma_d->conf_data->priv_conf[0]) {
 			EMSG("hpdma channel priv conf is incorrect");
 			panic();
 		}
 
 		if ((io_read32(hpdma_d->base + _HPDMA_SECCFGR) &
-		     hpdma_d->conf_data.access_mask[0]) !=
-		    hpdma_d->conf_data.sec_conf[0]) {
+		     hpdma_d->conf_data->access_mask[0]) !=
+		    hpdma_d->conf_data->sec_conf[0]) {
 			EMSG("hpdma channel sec conf is incorrect");
 			panic();
 		}
@@ -251,27 +254,33 @@ static TEE_Result parse_dt(const void *fdt, int node,
 		return res;
 
 	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
-	if (!cuint)
-		panic("No RIF configuration available");
+	if (!cuint) {
+		DMSG("No RIF configuration available");
+		return TEE_SUCCESS;
+	}
+
+	hpdma_d->conf_data = calloc(1, sizeof(*hpdma_d->conf_data));
+	if (!hpdma_d->conf_data)
+		panic();
 
 	hpdma_d->nb_channels = (unsigned int)(lenp / sizeof(uint32_t));
 	assert(hpdma_d->nb_channels <= HPDMA_RIF_CHANNELS);
 
-	hpdma_d->conf_data.cid_confs = calloc(HPDMA_RIF_CHANNELS,
-					      sizeof(uint32_t));
-	hpdma_d->conf_data.sec_conf = calloc(1, sizeof(uint32_t));
-	hpdma_d->conf_data.priv_conf = calloc(1, sizeof(uint32_t));
-	hpdma_d->conf_data.access_mask = calloc(1, sizeof(uint32_t));
-	hpdma_d->conf_data.lock_conf = calloc(1, sizeof(uint32_t));
-	if (!hpdma_d->conf_data.cid_confs || !hpdma_d->conf_data.sec_conf ||
-	    !hpdma_d->conf_data.priv_conf || !hpdma_d->conf_data.access_mask ||
-	    !hpdma_d->conf_data.lock_conf)
+	hpdma_d->conf_data->cid_confs = calloc(HPDMA_RIF_CHANNELS,
+					       sizeof(uint32_t));
+	hpdma_d->conf_data->sec_conf = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data->priv_conf = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data->access_mask = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data->lock_conf = calloc(1, sizeof(uint32_t));
+	if (!hpdma_d->conf_data->cid_confs || !hpdma_d->conf_data->sec_conf ||
+	    !hpdma_d->conf_data->priv_conf ||
+	    !hpdma_d->conf_data->access_mask || !hpdma_d->conf_data->lock_conf)
 		panic("Missing memory capacity for HPDMA RIF configuration");
 
 	for (i = 0; i < hpdma_d->nb_channels; i++) {
 		rif_conf = fdt32_to_cpu(cuint[i]);
 
-		stm32_rif_parse_cfg(rif_conf, &hpdma_d->conf_data,
+		stm32_rif_parse_cfg(rif_conf, hpdma_d->conf_data,
 				    HPDMA_NB_MAX_CID_SUPPORTED,
 				    HPDMA_RIF_CHANNELS);
 	}
@@ -295,25 +304,26 @@ static void stm32_hpdma_pm_suspend(struct hpdma_pdata *hpdma)
 	size_t i = 0;
 
 	for (i = 0; i < HPDMA_RIF_CHANNELS; i++)
-		hpdma->conf_data.cid_confs[i] = io_read32(hpdma->base +
-							  _HPDMA_CIDCFGR(i)) &
-						_HPDMA_CIDCFGR_CONF_MASK;
-
-	hpdma->conf_data.priv_conf[0] = io_read32(hpdma->base +
-						  _HPDMA_PRIVCFGR) &
-					_HPDMA_PRIVCFGR_MASK;
-	hpdma->conf_data.sec_conf[0] = io_read32(hpdma->base +
-						 _HPDMA_SECCFGR) &
-				       _HPDMA_SECCFGR_MASK;
-	hpdma->conf_data.lock_conf[0] = io_read32(hpdma->base +
-						  _HPDMA_RCFGLOCKR) &
-					_HPDMA_RCFGLOCKR_MASK;
+		hpdma->conf_data->cid_confs[i] = io_read32(hpdma->base +
+							   _HPDMA_CIDCFGR(i)) &
+						 _HPDMA_CIDCFGR_CONF_MASK;
+
+	hpdma->conf_data->priv_conf[0] = io_read32(hpdma->base +
+						   _HPDMA_PRIVCFGR) &
+					 _HPDMA_PRIVCFGR_MASK;
+	hpdma->conf_data->sec_conf[0] = io_read32(hpdma->base +
+						  _HPDMA_SECCFGR) &
+					_HPDMA_SECCFGR_MASK;
+	hpdma->conf_data->lock_conf[0] = io_read32(hpdma->base +
+						   _HPDMA_RCFGLOCKR) &
+					 _HPDMA_RCFGLOCKR_MASK;
 
 	/*
 	 * The access mask is modified to restore the conf for all
 	 * resources.
 	 */
-	hpdma->conf_data.access_mask[0] = GENMASK_32(HPDMA_RIF_CHANNELS - 1, 0);
+	hpdma->conf_data->access_mask[0] = GENMASK_32(HPDMA_RIF_CHANNELS - 1,
+						      0);
 }
 
 static TEE_Result
@@ -362,6 +372,13 @@ static TEE_Result stm32_hpdma_probe(const void *fdt, int node,
 
 	res = parse_dt(fdt, node, hpdma_d);
 	if (res) {
+		if (hpdma_d->conf_data) {
+			free(hpdma_d->conf_data->access_mask);
+			free(hpdma_d->conf_data->cid_confs);
+			free(hpdma_d->conf_data->priv_conf);
+			free(hpdma_d->conf_data->sec_conf);
+		}
+		free(hpdma_d->conf_data);
 		free(hpdma_d);
 		return res;
 	}
-- 
2.47.2

